File: bff-api/internal/services/contact_timeline_service.go
package services

import (
	"encoding/csv"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/maxiking445/bff-api/internal/common"
	"github.com/maxiking445/bff-api/internal/models"
)

func BuildContactTimelinesFromCSV(
	csvPath string,
) ([]models.ContactTimeline, error) {

	fmt.Println("Analyse Groupe File: " + csvPath)
	rows, err := loadContactMessagesFromCSV(csvPath)
	if err != nil {
		return nil, err
	}

	// Map[user]map[date]count
	userDayCounts := make(map[string]map[string]int)

	for _, msg := range rows {
		user := msg.Identity
		day := msg.Timestamp.Format("2006-01-02")

		if _, ok := userDayCounts[user]; !ok {
			userDayCounts[user] = make(map[string]int)
		}
		userDayCounts[user][day]++
	}
	// Map in []GroupTimeline
	var timelines []models.ContactTimeline
	for user, dayMap := range userDayCounts {
		var timeline []models.DayCount
		for day, count := range dayMap {
			timeline = append(timeline, models.DayCount{
				Date:  day,
				Count: count,
			})
		}

		sort.Slice(timeline, func(i, j int) bool {
			return timeline[i].Date < timeline[j].Date
		})
		if user == "" {
			user = "You"
		}
		println(user)
		timelines = append(timelines, models.ContactTimeline{
			User:     LoadIdentityByUserUUID(user),
			Timeline: timeline,
		})
	}

	return timelines, nil
}

type ContactMessageRow struct {
	Identity  string
	Timestamp time.Time
}

func loadContactMessagesFromCSV(path string) ([]ContactMessageRow, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	reader := csv.NewReader(f)
	records, err := reader.ReadAll()
	if err != nil {
		return nil, err
	}

	if len(records) <= 1 {
		return []ContactMessageRow{}, nil
	}

	var rows []ContactMessageRow
	for i, r := range records {
		if i == 0 {
			continue
		}

		base := filepath.Base(path)
		base = strings.TrimSuffix(base, common.CSV)
		identity := strings.TrimPrefix(base, common.UserMessage)
		createdAtMillis, err := strconv.ParseInt(r[7], 10, 64)
		if err != nil {
			fmt.Println("Error loading created_at:", r[7], "Error:", err)
			continue
		}
		timestamp := time.UnixMilli(createdAtMillis)
		if r[5] == "READ" {
			identity = ""
		}

		rows = append(rows, ContactMessageRow{
			Identity:  identity,  // identity = user
			Timestamp: timestamp, //Timestamp
		})
	}

	return rows, nil
}

File: analyzer-frontend/src/service/ContactTimelineService.ts
import Papa from "papaparse";
import { getZipFile } from "./FileService";
import { loadIdentityByUserUUID } from "./IdentityService";
import { ModelsContactTimeline } from "@/models/ModelsContactTimeline";
import { ModelsDayCount } from "@/models/ModelsDayCount";

interface ContactMessageRow {
  identity: string;
  timestamp: Date;
}

export async function getContactTimeline(userId: string): Promise<ModelsContactTimeline[]> {
  if (!userId) throw new Error("contact missing");

  const identity = await loadIdentityByUserUUID(userId);
  if (!identity.identityID) throw new Error(`Identity not found for userId: ${userId}`);

  const filename = `message_${identity.identityID}.csv`;
  return await buildContactTimelinesFromCSV(filename);
}

async function loadContactMessagesFromCSV(
  filename: string,
): Promise<ContactMessageRow[]> {
  const file = getZipFile(filename);
  if (!file) return [];

  const text = await file.content.text();
  const records = Papa.parse(text, { header: false, skipEmptyLines: true })
    .data as string[][];

  if (records.length <= 1) return [];

  const base = filename.replace(/^message_/, "").replace(/\.csv$/, "");
  const rows: ContactMessageRow[] = [];

  for (const row of records.slice(1)) {
    const createdAtMillis = parseInt(row[7], 10);
    if (isNaN(createdAtMillis)) continue;

    const identity = row[5] === "READ" ? "" : base;

    rows.push({
      identity,
      timestamp: new Date(createdAtMillis),
    });
  }

  return rows;
}

export async function buildContactTimelinesFromCSV(
  filename: string,
): Promise<ModelsContactTimeline[]> {
  const rows = await loadContactMessagesFromCSV(filename);

  // Map[user][date] = count
  const userDayCounts = new Map<string, Map<string, number>>();

  for (const msg of rows) {
    const user = msg.identity || "You";
    const day = msg.timestamp.toISOString().split("T")[0];

    if (!userDayCounts.has(user)) {
      userDayCounts.set(user, new Map());
    }
    const dayMap = userDayCounts.get(user)!;
    dayMap.set(day, (dayMap.get(day) ?? 0) + 1);
  }

  const timelines: ModelsContactTimeline[] = [];

  for (const [user, dayMap] of userDayCounts) {
    const timeline: ModelsDayCount[] = Array.from(dayMap.entries())
      .map(([date, count]) => ({ date, count }))
      .sort((a, b) => (a.date ?? "").localeCompare(b.date ?? ""));

    timelines.push({
      identity: await loadIdentityByUserUUID(user),
      timeline,
    });
  }

  return timelines;
}